lib/Doctrine/DBAL/Driver/ResultStatement.php

```diff
@@ -9,6 +9,59 @@
  */
 interface ResultStatement extends \Traversable
 {
+    /**
+     * Specifies that the fetch method shall return each row as an array indexed
+     * by column name as returned in the corresponding result set. If the result
+     * set contains multiple columns with the same name, the statement returns
+     * only a single value per column name.
+     *
+     * @see \PDO::FETCH_ASSOC
+     */
+    const FETCH_ASSOC = 2;
```

Comment:
I'd prefer moving these out to FetchMode enum-like class, pretty much like ORM 3.x will have i.e. GeneratorType. Either as a simple constant holder (ORM case ATM) or proper object encapsulation (encapsulates enum-like validation).
Also this is the opportunity to fix the naming - I'm not a fan of abbreviations tbh.

Reply:
What's the benefit of having an enum-class with constants? Right now, `FETCH_*`, `PARAM_*` and `CASE_*` constants belong to the interfaces which implement their handling which means better cohesion. Do you want to introduce an enum-class for each group of constants?

Reply:
> What's the benefit of having an enum-class with constants?

To me, it seems more consistent, because:
1) they're grouped only by their purpose,
2) they're not mixed with other unrelated constants,
3) their naming is more comprehensive,
3) they could be typed if enum-like structure is used (something like [this](https://github.com/consistence/consistence/blob/master/docs/Enum/enums.md), possibly with better/simpler API)

> Could you draft the non-abbreviated constant names as you see them?

In this case something like this:
```php
class FetchType
{
    public const ASSOCIATIVE = 0b1;
    public const NUMERIC     = 0b01;
    public const COLUMN      = 0b001;
    public const OBJECT      = 0b0001;
    public const MIXED       = self::ASSOCIATIVE | self::NUMERIC;
}
```
```php
class ParameterFormat
{
    public const NULL         = 0b1;
    public const INTEGER      = 0b01;
    public const STRING       = 0b001;
    public const BOOLEAN      = 0b0001;
    public const LARGE_OBJECT = 0b00001;
}
```
(Don't take it as a proposal to change here, rather something to consider for 3.x.)

Reply:
> Don't take it as a proposal to change here, rather something to consider for 3.x.

The approach currently taken is to introduce the new constants in `2.x` and remove the assignment of the `PDO` values to them in `3.x`. This way, the callers will have a period of time when either of the APIs can be used.

Not having this changes introduced in `2.x` will make the transition much more painful.

As for the actual values, should they really be bits? It will create an impression that they can be mixed together however they are mutually exclusive. To me, a set of `1`, `2`, `3`… would look more natural. Otherwise, what if somebody calls `$stmt->fetch(FetchType::COLUMN | FetchType::OBJECT)`? We'll have to have a validation layer for those bit masks which could be avoided.

Reply:
`ParameterFormat` should be probably `ParameterType`. Formatting implies a string representation of some value (like `sprintf()`). Here, we're not dealing with any formatting.

Reply:
`FetchType` — personally, I'd prefer to keep it as `FetchMode`. The "mode" implies "the way of doing things" which is exactly what those constants define. The "type" is too generic IMO.

Reply:
> `public const MIXED = self::ASSOCIATIVE | self::NUMERIC;`

Since it's being brought up, does the DBAL need to support this weird mode, especially defined this way? In my entire experience, I have never needed to fetch the data this way. I think it's the same sort of feature as register globals in PHP 4 which allowed developers to do both `$row[0]` and `$row['id']` without really thinking of what the `$row` actually is.